processing/src/main/java/io/druid/query/QueryRunnerFactory.java

```diff
@@ -57,7 +57,7 @@
    * @param queryRunners Individual QueryRunner objects that produce some results
    * @return a QueryRunner that, when asked, will use the ExecutorService to runt he base QueryRunners
    */
-  public QueryRunner<T> mergeRunners(ExecutorService queryExecutor, Iterable<QueryRunner<T>> queryRunners);
+  public QueryRunner<T> mergeRunners(ExecutorService queryExecutor, Iterable<QueryRunner<T>> queryRunners, QueryType query);
```

Comment:
one possible alternative could be to move the single threaded vs multi threaded logic from QueryRunnerFactory to GroupByParallelQueryRunner, which can look at the context and choose whether to do single/multi threaded run depending on the context.